main/src/cgeo/geocaching/Image.java

```diff
@@ -15,27 +17,92 @@
 
 import java.io.File;
 
+/**
+ * Represent an Image along with Title and Description.
+ * Mostly used for representing image in/from logs.
+ */
 public class Image implements Parcelable {
-    private final String url;
-    private final String title;
-    private final String description;
+    @NonNull private Uri uri;
```

Comment:
Good link. The approved answer to this Stack Exchange question summarizes my point very well:

> Apart from all these legitimate reasons, alas, the most probable cause why people keep creating mutable objects is inertia of mind, a.k.a. resistance to change. Note that most developers of today have been trained well before immutability (and the containing paradigm, functional programming) became "trendy" in their sphere of influence, and don't keep their knowledge up to date about new tools and methods of our trade - in fact, many of us humans positively resist new ideas and processes. "I have been programming like this for nn years and I don't care about the latest stupid fads!"

Ideally, we should have a static `Image.builder()` create an mutable object, with setters and a `build()` method which then creates an immutable `Image`. Unfortunately, doing so is a very non-interesting-to-code task, but leads to much better code. Once you have an immutable object, coding `equals()` and `hashCode()` is trivial and safe (the fields never change, so the equality results and the hash-code don't either, which makes it safe to include as keys in a map for example).